stdlib/public/core/CollectionAlgorithms.swift.gyb

```diff
@@ -142,149 +142,84 @@ orderingExplanation = """\
 // partition()
 //===----------------------------------------------------------------------===//
 
-% # Generate two versions: with explicit predicates and with
-% # a Comparable requirement.
-% for preds in [True, False]:
-
-%   if preds:
-
-extension MutableCollection where Self : RandomAccessCollection
-
-%   else:
-
-extension MutableCollection
-  where Self : RandomAccessCollection, ${IElement} : Comparable
-
-%   end
-{
-
-%   if preds:
-  /// Reorders the elements in the collection and returns a pivot index, using
-  /// the given predicate as the comparison between elements.
-  ///
-  /// This method is typically one step of a sorting algorithm. A collection is
-  /// partitioned around a pivot index when each of the elements before the
-  /// pivot is correctly ordered before each of the elements at or after the
-  /// pivot. The `partition(isOrderedBefore:)` method reorders the elements of
-  /// the collection and returns a pivot index that satisfies this condition,
-  /// using the given predicate to determine the relative order of any two
-  /// elements.
-  ///
-${orderingExplanation}
-  /// Here's an example that uses a predicate that orders elements from largest
-  /// to smallest:
-  ///
-  ///     var numbers = [50, 30, 60, 50, 80, 10, 40, 30]
-  ///     let pivot = numbers.partition { a, b in a > b }
-  ///
-  ///     print(pivot)
-  ///     // Prints "2"
-  ///     print(numbers)
-  ///     // Prints "[60, 80, 50, 50, 30, 10, 40, 30]"
-  ///
-  /// The return value of the call to `numbers.partition()` is the pivot for
-  /// the rearranged `numbers` array. `pivot` divides the collection into two
-  /// subranges, `numbers[0..<pivot]` and `numbers[pivot..<8]`.
-  ///
-  ///     print(numbers[0..<pivot])
-  ///     // Prints "[60, 80]"
-  ///     print(numbers[pivot..<8])
-  ///     // Prints "[50, 50, 30, 10, 40, 30]"
-  ///
-  /// The elements of `numbers` are rearranged so that every element in the
-  /// subrange before `pivot` is ordered before every element in the subrange
-  /// after. Because the supplied predicate returns `true` when its first
-  /// argument is greater than its second argument, larger elements are
-  /// ordered before smaller elements.
-  ///
-  /// - Parameter isOrderedBefore: A predicate that returns `true` if its first
-  ///   argument should be ordered before its second argument; otherwise,
-  ///   `false`.
-  /// - Returns: A pivot index, such that every element before the pivot is
-  ///   ordered before every element at or above the pivot, using
-  ///   `isOrderedBefore` to determine the relative order of any two elements.
-  ///   The returned pivot is equal to the collection's end index only if the
-  ///   collection is empty.
-  ///
-  /// - SeeAlso: `partition()`
+extension MutableCollection {
   public mutating func partition(
-    isOrderedBefore: @noescape (${IElement}, ${IElement}) -> Bool
-  ) -> Index
+      by belongsInSecondPartition: @noescape (${IElement}) throws -> Bool
```

Comment:
Actually, since you have different implementations now, you should make `partition()` a requirement of `MutableCollection` so that we get the most efficient implementation via dynamic dispatch.  Then you'd need just one test for forward collections, it will dispatch to both implementations depending on the type.